File: src/main/cpp/lib/ReportGenerator.cpp
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 *******************************************************************************/

#include "report/ReportGenerator.h"
#include "storage/Repository.h"
#include "purchase/Invoice.h"
#include "purchase/PurchasedBook.h"

using namespace purchase;

namespace report
{

	ReportGenerator::ReportGenerator() : repository_(main_repository::configuredRepository())
	{
	}

	double ReportGenerator::getTotalAmount() const
	{
		const auto invoiceMap = repository_->getInvoiceMap();
		double totalAmount = 0.0;
		for (const auto id2Invoice : invoiceMap)
		{
			totalAmount += id2Invoice.second->computeTotalAmount();
		}
		return totalAmount;
	}

	int ReportGenerator::getTotalSoldBooks() const
	{
		const auto invoiceMap = repository_->getInvoiceMap();
		int totalSoldBooks = 0;
		for (const auto id2Invoice : invoiceMap)
		{
			for (const auto purchasedBook : id2Invoice.second->getPurchasedBooks())
			{
				totalSoldBooks += purchasedBook->getQuantity();
			}
		}
		return totalSoldBooks;
	}

	size_t ReportGenerator::getNumberOfIssuedInvoices() const
	{
		const auto invoiceMap = repository_->getInvoiceMap();
		return invoiceMap.size();
	}
}

File: src/main/csharp/Application/Report/ReportGenerator.cs
using System.Collections.Generic;
using System.Linq;
using Application.Purchase;
using Application.Storage;

namespace Application.Report
{
    public class ReportGenerator
    {
        private readonly IRepository _repository = MainRepository.ConfiguredRepository;

        public double GetTotalAmount()
        {
            var invoices = _repository.GetInvoiceMap().Values;
            var totalAmount = invoices.Sum(invoice => invoice.ComputeTotalAmount());
            return totalAmount;
        }

        public int GetTotalSoldBooks()
        {
            var invoices = _repository.GetInvoiceMap().Values;
            var totalSoldBooks = invoices
                .Sum(invoice => invoice.PurchasedBooks.Sum(purchasedBook => purchasedBook.Quantity));

            return totalSoldBooks;
        }

        public long GetNumberOfIssuedInvoices() => _repository.GetInvoiceMap().Count;
    }
}